### PR TITLE
BulkData $import API improvement - Exit Status #949

### PR DESCRIPTION
**Pull request by prb112**
- Updated the ChunkJob.xml to include missing parameters (the passed
values didn't agree)
- Add guard for null condition
- Added support for operational outcome bucket
```
{
    "transactionTime": "2020-04-25T20:35:50.699-04:00",
    "request": "$import",
    "requiresAccessToken": false,
    "output": [
        {
            "type": "OperationOutcome",
            "url": "test-import.ndjson_oo_success.ndjson",
            "count": 3
        }
    ],
    "error": [
        {
            "type": "OperationOutcome",
            "url": "test-import.ndjson_oo_errors.ndjson",
            "count": 0
        }
    ]
}```

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkImportChunkJob.xml [hunk 1]
```diff
@@ -49,12 +49,13 @@
                     <property name="cos.location" value="#{jobParameters['cos.location']}"/>
                     <property name="cos.credential.ibm" value="#{jobParameters['cos.credential.ibm']}"/>
                     <property name="cos.bucket.name" value="#{jobParameters['cos.bucket.name']}"/>
+                    <property name="cos.operationoutcomes.bucket.name" value="#{jobParameters['cos.operationoutcomes.bucket.name']}"/>
```
Comment (albertwang-ibm) [inline fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkImportChunkJob.xml:52]: Mapper doesn't need this.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionCollector.java [hunk 2]
```diff
@@ -80,10 +86,7 @@ public ImportPartitionCollector() throws Exception {
                 logger.finer("ImportPartitionCollector: Succeed get CosClient!");
             }
         }
-    }
 
-    @Override
-    public Serializable collectPartitionData() throws Exception{
```
Comment (albertwang-ibm) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionCollector.java:86]: not sure why this change is needed, not inject property is used in ImportPartitionCollector(), the CosClient was created in ImportPartitionCollector() to make sure it's only created once for each ImportPartitionCollector object.

Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionCollector.java:86]: it was null, always null when executed

Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionCollector.java:86]: I debugged it, and found it was injected after object creation

File: fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkImportChunkJob.xml (hunk 1 above)
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkImportChunkJob.xml:52]:
```suggestion
```

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java [hunk 3]
```diff
@@ -269,6 +276,13 @@ public PartitionPlan mapPartitions() throws Exception {
             Properties p = new Properties();
             p.setProperty(Constants.IMPORT_PARTITTION_WORKITEM, fhirDataSource.getUrl());
             p.setProperty(Constants.IMPORT_PARTITTION_RESOURCE_TYPE, fhirDataSource.getType());
+            p.setProperty(Constants.COS_API_KEY, cosApiKeyProperty);
+            p.setProperty(Constants.COS_SRVINST_ID, cosSrvinstId);
+            p.setProperty(Constants.COS_ENDPOINT_URL, cosEndpointUrl);
+            p.setProperty(Constants.COS_LOCATION, cosLocation);
+            p.setProperty(Constants.COS_OPERATIONOUTCOMES_BUCKET_NAME, cosOperationOutcomesBucketName);
+            p.setProperty(Constants.COS_IS_IBM_CREDENTIAL, cosCredentialIbm);
+
```
Comment (albertwang-ibm) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java:285]: Mapper already got these properties from injection, don't need to pass through partition properties.

File: fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkImportChunkJob.xml (hunk 1 above)
Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkImportChunkJob.xml]: Right, fixed

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java (hunk 3 above)
Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java:285]: That's not what I found, I found the properties were empty when it went to the step execution

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/client/BulkDataClient.java [hunk 4]
```diff
@@ -448,6 +453,7 @@ public String submitImport(String inputFormat, String inputSource, List<Input> i
         builder.applicationName(properties.get(BulkDataConfigUtil.APPLICATION_NAME));
         builder.moduleName(properties.get(BulkDataConfigUtil.MODULE_NAME));
         builder.cosBucketName(properties.get(BulkDataConfigUtil.JOB_PARAMETERS_BUCKET));
+        builder.cosBucketNameOo(properties.get(BulkDataConfigUtil.JOB_PARAMETERS_BUCKET));
```
Comment (albertwang-ibm) [inline fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/client/BulkDataClient.java:456]: better use OperationOutcome instead of Oo?

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobExecutionResponse.java [hunk 5]
```diff
@@ -340,6 +340,11 @@ public Builder fhirStorageType(StorageDetail storageDetails) {
             jobParameter.setStorageDetails(storageDetails);
             return this;
         }
+
+        public Builder cosBucketNameOo(String cosBucketNameOo) {
```
Comment (albertwang-ibm) [inline fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobExecutionResponse.java:344]: better use OperationOutcome instead of Oo?

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobInstanceRequest.java [hunk 6]
```diff
@@ -190,6 +190,11 @@ public JobInstanceRequest build() {
             request.setJobParameters(jobParameter);
             return request;
         }
+
+        public Builder cosBucketNameOo(String cosBucketNameOo) {
```
Comment (albertwang-ibm) [inline fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobInstanceRequest.java:194]: Oo seems strange ...

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/client/BulkDataClient.java (hunk 4 above)
Comment (prb112) [reply fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/client/BulkDataClient.java:456]: Fixed in next commit

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobExecutionResponse.java (hunk 5 above)
Comment (prb112) [reply fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobExecutionResponse.java:344]: Fixed in next commit

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobInstanceRequest.java (hunk 6 above)
Comment (prb112) [reply fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobInstanceRequest.java:194]: Fixed in next commit

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/client/BulkDataClient.java (hunk 4 above)
Comment (prb112) [reply fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/client/BulkDataClient.java:456]: Fixed in next commit

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobExecutionResponse.java (hunk 5 above)
Comment (prb112) [reply fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobExecutionResponse.java:344]: Fixed in next commit

File: fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobInstanceRequest.java (hunk 6 above)
Comment (prb112) [reply fhir-operation-bulkdata/src/main/java/com/ibm/fhir/operation/bulkdata/model/JobInstanceRequest.java:194]: Fixed in next commit

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java (hunk 3 above)
Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java:285]: backed out this change.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionCollector.java (hunk 2 above)
Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionCollector.java:86]: I am leaving this code as-is

Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionCollector.java:86]: en... this is ugly. so this means the constructor is never called.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java (hunk 3 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java:272]: I don't really think this is correct, I mean if so, the import job should have always failed.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionCollector.java (hunk 2 above)
Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionCollector.java:86]: The constructor is called... then injection... you're depending on the injection to happen prior to the constructor call .

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java (hunk 3 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java:272]: sorry, I mean the chunkjob steps already got these properties from injection. don't need to be set by the mapper.  otherwise, the import job should have never succeeded from the beginning ...

Comment (prb112) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkimport/ImportPartitionMapper.java:272]: Yes, no worries, it was confused with the Collector problem. All fixed now

**Comment by albertwang-ibm** [comment]
I'm merging it and then sync to my local for my export refactoring.
